File: webui/src/lib/repos/seed.ts
import { UNCATEGORIZED_ID, type Category, type Expense } from '$lib/types';
import { colorForCategoryId, UNCATEGORIZED_COLOR } from '$lib/utils/categoryColor';

type CategorySeed = {
	id: string;
	name: string;
	icon: string;
	description?: string;
};

type ExpenseSeed = {
	id: `exp-${string}`;
	name: string;
	amount: number;
	categoryId: string;
	note: string;
	importance?: Expense['importance'];
	monthsAgo: number;
	day: number;
};

const CATEGORY_SEEDS: readonly CategorySeed[] = [
	{
		id: UNCATEGORIZED_ID,
		name: 'Uncategorized',
		icon: 'circle-help',
		description: 'Transactions that have not been assigned to any category yet.',
	},
	{
		id: 'cat-housing',
		name: 'Housing',
		icon: 'house',
		description:
			'Home-related expenses and utilities: rent, mortgage, electricity, water, gas, internet, maintenance, furniture, and household items.',
	},
	{
		id: 'cat-groceries',
		name: 'Groceries',
		icon: 'shopping-cart',
		description:
			'Food and household consumables bought for home use, including supermarkets, baby food, cleaning supplies, and toiletries bought with groceries. Do not use for restaurants, cafes, or delivery.',
	},
	{
		id: 'cat-health',
		name: 'Health',
		icon: 'heart-pulse',
		description:
			'Medical and healthcare spending: pharmacy, doctor, dental, vitamins, supplements, tests, and health-related products.',
	},
	{
		id: 'cat-childcare',
		name: 'Childcare',
		icon: 'baby',
		description:
			'Direct child care costs: nursery, kindergarten, nanny, babysitter, child activities, and school-related care. Do not use for toys, clothes, or baby food.',
	},
	{
		id: 'cat-car',
		name: 'Car',
		icon: 'car',
		description:
			'Car ownership and maintenance: fuel, insurance, parking, repairs, car wash, registration, and vehicle taxes.',
	},
	{
		id: 'cat-public-transport',
		name: 'Public Transport',
		icon: 'train-front',
		description:
			'Daily non-taxi transport: bus, metro, train, tram, and similar public transit fares.',
	},
	{
		id: 'cat-eating-out',
		name: 'Eating Out',
		icon: 'utensils',
		description:
			'Food and drinks bought outside the home: restaurants, cafes, coffee shops, food delivery, fast food, and outside snacks.',
	},
	{
		id: 'cat-taxi',
		name: 'Taxi',
		icon: 'car-taxi-front',
		description:
			'Taxi and ride-sharing convenience transport, including Uber, Bolt, local taxis, and similar services.',
	},
	{
		id: 'cat-shopping',
		name: 'Shopping',
		icon: 'shopping-bag',
		description:
			'General personal and household shopping: clothes, shoes, cosmetics, toys, small household purchases, and non-essential purchases. Do not use for technology purchases.',
	},
	{
		id: 'cat-technology-gadgets',
		name: 'Technology & Gadgets',
		icon: 'smartphone',
		description:
			'Technology purchases and accessories: phones, laptops, tablets, smart watches, smart home devices, headphones, and computer accessories.',
	},
	{
		id: 'cat-sports-fitness',
		name: 'Sports & Fitness',
		icon: 'dumbbell',
		description:
			'Fitness and sports spending: gym memberships, fitness classes, sporting equipment, running shoes, sports clubs, and fitness apps.',
	},
	{
		id: 'cat-entertainment-leisure',
		name: 'Entertainment & Leisure',
		icon: 'ticket',
		description:
			'Entertainment, hobbies, and leisure: cinema, concerts, video games, books, hobbies, events, and recreational activities. Do not use for recurring subscriptions.',
	},
	{
		id: 'cat-subscriptions',
		name: 'Subscriptions',
		icon: 'repeat',
		description:
			'Recurring digital or membership payments only: Netflix, Spotify, iCloud, YouTube Premium, software subscriptions, and Amazon Prime.',
	},
	{
		id: 'cat-travel',
		name: 'Travel',
		icon: 'plane',
		description:
			'Non-daily travel and vacation costs: flights, hotels, vacation transportation, travel activities, and travel bookings.',
	},
	{
		id: 'cat-gifts',
		name: 'Gifts',
		icon: 'gift',
		description:
			'Gifts and celebrations: birthday gifts, holiday gifts, flowers, celebration expenses, and special occasions.',
	},
];

const EXPENSE_SEEDS: readonly ExpenseSeed[] = [
	{ id: 'exp-001', name: 'Whole Foods', amount: 58.24, categoryId: 'cat-groceries', note: 'Weekly groceries', monthsAgo: 0, day: 3 },
	{ id: 'exp-002', name: 'Transport for London', amount: 14.5, categoryId: 'cat-public-transport', note: 'Bus fare', monthsAgo: 0, day: 7 },
	{ id: 'exp-003', name: 'Starbucks', amount: 42.8, categoryId: 'cat-eating-out', note: 'Coffee and lunch', monthsAgo: 1, day: 11 },
	{ id: 'exp-004', name: 'Greystar Rent', amount: 1200, categoryId: 'cat-housing', note: '', monthsAgo: 1, day: 1 },
	{ id: 'exp-005', name: 'Comcast Xfinity', amount: 96.12, categoryId: 'cat-housing', note: 'Internet bill', monthsAgo: 2, day: 5 },
	{ id: 'exp-006', name: 'Trader Joe’s', amount: 33.75, categoryId: 'cat-groceries', note: 'Pantry restock', monthsAgo: 2, day: 13 },
	{ id: 'exp-007', name: 'Uber', amount: 22, categoryId: 'cat-taxi', note: 'Rideshare', monthsAgo: 3, day: 9 },
	{ id: 'exp-008', name: 'Deliveroo', amount: 68.4, categoryId: 'cat-eating-out', note: 'Dinner out', monthsAgo: 3, day: 19 },
	{ id: 'exp-009', name: 'Sainsbury’s', amount: 48.9, categoryId: 'cat-groceries', note: 'Farmers market', monthsAgo: 4, day: 6 },
	{ id: 'exp-010', name: 'Pacific Gas & Electric', amount: 74.3, categoryId: 'cat-housing', note: 'Electricity', monthsAgo: 4, day: 16 },
	{ id: 'exp-011', name: 'NCP Parking', amount: 15.2, categoryId: 'cat-car', note: 'Parking', monthsAgo: 5, day: 4 },
	{ id: 'exp-012', name: 'Chipotle', amount: 89.95, categoryId: 'cat-eating-out', note: 'Team lunch', monthsAgo: 5, day: 10 },
	{ id: 'exp-013', name: 'Amazon Prime', amount: 27.6, categoryId: 'cat-groceries', note: 'Snacks', monthsAgo: 0, day: 21 },
	{ id: 'exp-014', name: 'Thames Water', amount: 110.4, categoryId: 'cat-housing', note: 'Water bill', monthsAgo: 2, day: 24 },
	{ id: 'exp-015', name: 'IKEA', amount: 210.85, categoryId: 'cat-housing', note: 'Maintenance supplies', monthsAgo: 3, day: 27 },
	{ id: 'exp-016', name: 'Patreon', amount: 12, categoryId: UNCATEGORIZED_ID, note: 'Monthly membership', monthsAgo: 1, day: 14 },
	{ id: 'exp-017', name: 'Netflix', amount: 17.99, categoryId: 'cat-subscriptions', note: 'Streaming', monthsAgo: 0, day: 18 },
];

function pad(value: number): string {
	return String(value).padStart(2, '0');
}

function isoDateForMonthsAgo(reference: Date, monthsAgo: number, day: number): string {
	const date = new Date(reference.getFullYear(), reference.getMonth() - monthsAgo, day);
	return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}

export function defaultCategories(): Category[] {
	return CATEGORY_SEEDS.map((seed) => ({
		id: seed.id,
		name: seed.name,
		color: seed.id === UNCATEGORIZED_ID ? UNCATEGORIZED_COLOR : colorForCategoryId(seed.id),
		icon: seed.icon,
		description: seed.description,
	}));
}

export function sampleExpenses(): Expense[] {
	const reference = new Date();
	return EXPENSE_SEEDS.map((seed) => ({
		id: seed.id,
		name: seed.name,
		amount: seed.amount,
		date: isoDateForMonthsAgo(reference, seed.monthsAgo, seed.day),
		categoryId: seed.categoryId,
		note: seed.note,
		importance: seed.importance ?? 'important',
	}));
}

File: api/src/quid_api/seed.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from decimal import Decimal
from typing import TYPE_CHECKING

from quid_api.category_helpers import UNCATEGORIZED_ID, color_for_category_id
from quid_api.models import Category, Expense

if TYPE_CHECKING:
    from quid_api.repositories.categories import CategoryRepository
    from quid_api.repositories.expenses import ExpenseRepository


@dataclass(frozen=True)
class _CatSeed:
    id: str
    name: str
    icon: str
    description: str


@dataclass(frozen=True)
class _ExpSeed:
    id: str
    name: str
    amount: str
    category_id: str
    note: str
    months_ago: int
    day: int


CATEGORY_SEEDS: tuple[_CatSeed, ...] = (
    _CatSeed(
        "cat-housing",
        "Housing",
        "house",
        "Home-related expenses and utilities: rent, mortgage, electricity, water, gas, internet, maintenance, furniture, and household items.",
    ),
    _CatSeed(
        "cat-groceries",
        "Groceries",
        "shopping-cart",
        "Food and household consumables bought for home use, including supermarkets, baby food, cleaning supplies, and toiletries bought with groceries. Do not use for restaurants, cafes, or delivery.",
    ),
    _CatSeed(
        "cat-health",
        "Health",
        "heart-pulse",
        "Medical and healthcare spending: pharmacy, doctor, dental, vitamins, supplements, tests, and health-related products.",
    ),
    _CatSeed(
        "cat-childcare",
        "Childcare",
        "baby",
        "Direct child care costs: nursery, kindergarten, nanny, babysitter, child activities, and school-related care. Do not use for toys, clothes, or baby food.",
    ),
    _CatSeed(
        "cat-car",
        "Car",
        "car",
        "Car ownership and maintenance: fuel, insurance, parking, repairs, car wash, registration, and vehicle taxes.",
    ),
    _CatSeed(
        "cat-public-transport",
        "Public Transport",
        "train-front",
        "Daily non-taxi transport: bus, metro, train, tram, and similar public transit fares.",
    ),
    _CatSeed(
        "cat-eating-out",
        "Eating Out",
        "utensils",
        "Food and drinks bought outside the home: restaurants, cafes, coffee shops, food delivery, fast food, and outside snacks.",
    ),
    _CatSeed(
        "cat-taxi",
        "Taxi",
        "car-taxi-front",
        "Taxi and ride-sharing convenience transport, including Uber, Bolt, local taxis, and similar services.",
    ),
    _CatSeed(
        "cat-shopping",
        "Shopping",
        "shopping-bag",
        "General personal and household shopping: clothes, shoes, cosmetics, toys, small household purchases, and non-essential purchases. Do not use for technology purchases.",
    ),
    _CatSeed(
        "cat-technology-gadgets",
        "Technology & Gadgets",
        "smartphone",
        "Technology purchases and accessories: phones, laptops, tablets, smart watches, smart home devices, headphones, and computer accessories.",
    ),
    _CatSeed(
        "cat-sports-fitness",
        "Sports & Fitness",
        "dumbbell",
        "Fitness and sports spending: gym memberships, fitness classes, sporting equipment, running shoes, sports clubs, and fitness apps.",
    ),
    _CatSeed(
        "cat-entertainment-leisure",
        "Entertainment & Leisure",
        "ticket",
        "Entertainment, hobbies, and leisure: cinema, concerts, video games, books, hobbies, events, and recreational activities. Do not use for recurring subscriptions.",
    ),
    _CatSeed(
        "cat-subscriptions",
        "Subscriptions",
        "repeat",
        "Recurring digital or membership payments only: Netflix, Spotify, iCloud, YouTube Premium, software subscriptions, and Amazon Prime.",
    ),
    _CatSeed(
        "cat-travel",
        "Travel",
        "plane",
        "Non-daily travel and vacation costs: flights, hotels, vacation transportation, travel activities, and travel bookings.",
    ),
    _CatSeed(
        "cat-gifts",
        "Gifts",
        "gift",
        "Gifts and celebrations: birthday gifts, holiday gifts, flowers, celebration expenses, and special occasions.",
    ),
)

EXPENSE_SEEDS: tuple[_ExpSeed, ...] = (
    _ExpSeed("exp-001", "Whole Foods", "58.24", "cat-groceries", "Weekly groceries", 0, 3),
    _ExpSeed("exp-002", "Transport for London", "14.50", "cat-public-transport", "Bus fare", 0, 7),
    _ExpSeed("exp-003", "Starbucks", "42.80", "cat-eating-out", "Coffee and lunch", 1, 11),
    _ExpSeed("exp-004", "Greystar Rent", "1200.00", "cat-housing", "", 1, 1),
    _ExpSeed("exp-005", "Comcast Xfinity", "96.12", "cat-housing", "Internet bill", 2, 5),
    _ExpSeed("exp-006", "Trader Joes", "33.75", "cat-groceries", "Pantry restock", 2, 13),
    _ExpSeed("exp-007", "Uber", "22.00", "cat-taxi", "Rideshare", 3, 9),
    _ExpSeed("exp-008", "Deliveroo", "68.40", "cat-eating-out", "Dinner out", 3, 19),
    _ExpSeed("exp-009", "Sainsbury's", "48.90", "cat-groceries", "Farmers market", 4, 6),
    _ExpSeed("exp-010", "Pacific Gas & Electric", "74.30", "cat-housing", "Electricity", 4, 16),
    _ExpSeed("exp-011", "NCP Parking", "15.20", "cat-car", "Parking", 5, 4),
    _ExpSeed("exp-012", "Chipotle", "89.95", "cat-eating-out", "Team lunch", 5, 10),
    _ExpSeed("exp-013", "Amazon Prime", "27.60", "cat-groceries", "Snacks", 0, 21),
    _ExpSeed("exp-014", "Thames Water", "110.40", "cat-housing", "Water bill", 2, 24),
    _ExpSeed("exp-015", "IKEA", "210.85", "cat-housing", "Maintenance supplies", 3, 27),
    _ExpSeed("exp-016", "Patreon", "12.00", UNCATEGORIZED_ID, "Monthly membership", 1, 14),
    _ExpSeed("exp-017", "Netflix", "17.99", "cat-subscriptions", "Streaming", 0, 18),
)


def _months_ago_date(reference: datetime, months_ago: int, day: int) -> str:
    year = reference.year
    month = reference.month - months_ago
    while month <= 0:
        month += 12
        year -= 1
    return f"{year:04d}-{month:02d}-{day:02d}"


async def seed_samples(
    cat_repo: CategoryRepository,
    exp_repo: ExpenseRepository,
    *,
    now: datetime | None = None,
) -> dict[str, int]:
    session = cat_repo.session
    reference = now or datetime.now(tz=UTC)
    cat_inserted = 0
    exp_inserted = 0

    for cat_seed in CATEGORY_SEEDS:
        existing_cat = await session.get(Category, cat_seed.id)
        if existing_cat is not None:
            continue
        cat = Category(
            id=cat_seed.id,
            name=cat_seed.name,
            color=color_for_category_id(cat_seed.id),
            icon=cat_seed.icon,
            description=cat_seed.description,
        )
        session.add(cat)
        cat_inserted += 1
    await session.flush()

    for exp_seed in EXPENSE_SEEDS:
        existing_exp = await session.get(Expense, exp_seed.id)
        if existing_exp is not None:
            continue
        exp = Expense(
            id=exp_seed.id,
            name=exp_seed.name,
            amount=Decimal(exp_seed.amount),
            date=_months_ago_date(reference, exp_seed.months_ago, exp_seed.day),
            category_id=exp_seed.category_id,
            note=exp_seed.note,
        )
        session.add(exp)
        exp_inserted += 1
    await session.flush()

    return {"categories": cat_inserted, "expenses": exp_inserted}


async def reset_and_seed(
    cat_repo: CategoryRepository,
    exp_repo: ExpenseRepository,
) -> dict[str, int]:
    from sqlalchemy import text

    session = cat_repo.session
    await session.execute(text("DELETE FROM import_rules"))
    await session.execute(text("DELETE FROM expenses"))
    await session.execute(
        text("DELETE FROM categories WHERE id != :uid"), {"uid": UNCATEGORIZED_ID}
    )
    await session.flush()
    return await seed_samples(cat_repo, exp_repo)
